src/helpers/getPSP22Balances.ts

```diff
@@ -0,0 +1,192 @@
+        address,
+      )
+      const balanceObj = _balanceOf?.output?.toPrimitive() as Record<string, any>
+
+      balance = new BN(
+        _balanceOf.output ? (balanceObj.ok as string) || (balanceObj.Ok as string) : '0',
+      )
+
+      const data = {
+        tokenSlug: slug,
+        tokenDecimals: decimals,
+        tokenSymbol: symbol,
+        balance,
+        iconPath,
+      }
+
+      const psp22BalanceData = parsePSP22Balance(data, formatterOptions)
+      return psp22BalanceData
+    }),
+  )
+
+  return result
+}
+
+/**
+ * Watches the PSP-22 token balances of the given `address` and returns it in a callback.
+ * The returned void function can be used to unsubscribe.
+ */
+export const watchPSP22Balances = (
+  api: ApiPromise,
+  address: string | AccountId | undefined,
+  callback: (data: PSP22BalanceData[]) => void,
+  formatterOptions?: BalanceFormatterOptions,
+): VoidFunction | null => {
+  const tokens = psp22Asset as Record<string, ChainAsset>
+  const psp22ContractMap: Record<string, ContractPromise> = {}
+
+  Object.entries(tokens).forEach(([slug, tokenInfo]) => {
+    psp22ContractMap[slug] = new ContractPromise(
+      api,
+      _PSP22_ABI,
+      tokenInfo.metadata?.contractAddress,
+    )
+  })
+
+  if (!address) {
+    const result = Object.values(tokens).map(({ slug, decimals, symbol, iconPath }) => {
+      return {
+        tokenSlug: slug,
+        tokenDecimals: decimals,
+        tokenSymbol: symbol,
+        iconPath,
+      }
+    })
+    callback(result)
+    return null
+  }
+
+  // Function to query the chain, parse data, and return promisified data
+  const fetchTokenBalances = async () =>
+    callback(
+      await Promise.all(
+        Object.values(tokens).map(async ({ slug, decimals, symbol, iconPath }) => {
+          let balance = new BN(0)
+
+          const contract = psp22ContractMap[slug]
+          const _balanceOf = await contract.query['psp22::balanceOf'](
+            address,
+            { gasLimit: getMaxGasLimit(api) },
+            address,
+          )
+          const balanceObj = _balanceOf?.output?.toPrimitive() as Record<string, any>
+
+          balance = new BN(
+            _balanceOf.output ? (balanceObj.ok as string) || (balanceObj.Ok as string) : '0',
+          )
+
+          const data = {
+            tokenSlug: slug,
+            tokenDecimals: decimals,
+            tokenSymbol: symbol,
+            balance,
+            iconPath,
+          }
+
+          const psp22BalanceData = parsePSP22Balance(data, formatterOptions)
+          return psp22BalanceData
+        }),
+      ),
+    )
+
+  fetchTokenBalances()
+
+  // Create intervalId which can be used to unsubscribe
+  const intervalId = setInterval(fetchTokenBalances, PSP22_TOKEN_BALANCE_SUBSCRIPTION_INTERVAL)
+
+  return () => {
+    clearInterval(intervalId)
+  }
+}
+
+/**
+ * Helper to parse the fetched PSP22 token balance data.
+ */
+
+export const parsePSP22Balance = (
```

Comment:
This function is a bit odd, it takes the same data (`PSP22BalanceData`) as it returns (`PSP22BalanceData`). Just one w/o balanceFormatted and one with which is is only implicitly noticeable...

Suggestion: Replace it `formatPSP22Balance` that takes `PSP22BalanceData` (and optionally `formatterOptions`) that **only** returns the `balanceFormatted` string which you then add to the `PSP22BalanceData` object outside of the function.

Reply:
Replaced in 7025175